### ISSUE TITLE
trying to build conda environment for offline install

### ISSUE THREAD

**Issue by tjwilli58** [psychopy/psychopy#2123]
I'm trying to build a conda environment for Psychopy for an offline computer. I did this successully with 1.90.3 this past summer, but I'm having trouble doing this for v3.0.0b12.

I first create an environment using a slightly modified environment-3.6.yml. The differences are that I comment out the dependency for `x264 `since that package doesn't exist for Windows, and I put `configobj` back in the dependencies list and comment out the [https://github.com/DiffSK/configobj/archive/master.zip](https://github.com/DiffSK/configobj/archive/master.zip) link in tie pip: section.

I first create the environment on a connected computer, so anaconda can download the needed packages, then do a `conda env list ---explicit` on that environment to get a list of the packages and I create a conda channel from that.  So far, so good. 

The tricky part is the `pip install` for Psychopy. I do a `pip download --no-cache` for each of the modules in the pip: section of the  enviroment file an put this in a folder for my wheels.

> - pip:
>   - arabic_reshaper
>   #- https://github.com/DiffSK/configobj/archive/master.zip
>   - esprima
>   - freetype-py
>   - pygame
>   - pyglet==1.3.0b1
>   - pyparallel
>   - python-gitlab
>   - sounddevice
>   - soundfile
> 

This is what I end up with:

> (base) C:\Users\Tim\Documents\Psychopy>dir wheels
>  Volume in drive C is Acer
>  Volume Serial Number is 1C3B-1723
> 
>  Directory of C:\Users\Tim\Documents\Psychopy\wheels
> 
> 11/30/2018  08:59 PM    <DIR>          .
> 11/30/2018  08:59 PM    <DIR>          ..
> 11/30/2018  08:57 PM            14,714 arabic_reshaper-2.0.14-py3-none-any.whl
> 11/30/2018  08:59 PM           154,154 certifi-2018.11.29-py2.py3-none-any.whl
> 11/30/2018  08:59 PM           166,898 cffi-1.11.5-cp36-cp36m-win_amd64.whl
> 11/30/2018  08:59 PM           133,356 chardet-3.0.4-py2.py3-none-any.whl
> 11/30/2018  08:57 PM            39,573 configparser-3.5.0.tar.gz
> 11/30/2018  08:59 PM            47,021 esprima-4.0.1.tar.gz
> 11/30/2018  08:59 PM           570,414 freetype_py-2.0.0.post6-py2.py3-none-win_amd64.whl
> 11/30/2018  08:57 PM           829,119 future-0.17.1.tar.gz
> 11/30/2018  08:59 PM            58,213 idna-2.7-py2.py3-none-any.whl
> 11/30/2018  08:59 PM           158,295 pycparser-2.19.tar.gz
> 11/30/2018  08:59 PM         4,242,265 pygame-1.9.4-cp36-cp36m-win_amd64.whl
> 11/30/2018  08:59 PM         1,010,371 pyglet-1.3.0b1-py2.py3-none-any.whl
> 11/30/2018  08:59 PM            38,836 pyparallel-0.2.2-py3-none-any.whl
> 11/30/2018  08:59 PM           110,201 python-gitlab-1.6.0.tar.gz
> 11/30/2018  08:59 PM            57,987 requests-2.20.1-py2.py3-none-any.whl
> 11/30/2018  08:59 PM            10,702 six-1.11.0-py2.py3-none-any.whl
> 11/30/2018  08:59 PM           167,731 sounddevice-0.3.12-py2.py3.cp26.cp27.cp32.cp33.cp34.cp35.cp36.cp37.pp27.pp32.pp33.pp34.pp35.pp36-none-win_amd64.whl
> 11/30/2018  08:59 PM           692,723 SoundFile-0.10.2-py2.py3.cp26.cp27.cp32.cp33.cp34.cp35.cp36.pp27.pp32.pp33-none-win_amd64.whl
> 11/30/2018  08:59 PM           118,086 urllib3-1.24.1-py2.py3-none-any.whl
>               19 File(s)      8,620,659 bytes
>                2 Dir(s)  137,991,757,824 bytes free
> 

If I do a `pip install -f wheels PsychoPy-3.0.0b12.tar.gz` on a networked machine, everything is fine and Psychopy gets installed, but if I try this on a non-network machine with
`pip install --no-index -f wheels PsychoPy-3.0.0b12.tar.gz`

It stops when the install gets to msgpack-python:

> (anaconda-py3.6) C:\Users\Tim\Documents\Psychopy>pip install --no-index -f wheels PsychoPy-3.0.0b12.tar.gz >installlog.txt
>   Could not find a version that satisfies the requirement msgpack-python (from PsychoPy==3.0.0b12) (from versions: )
> No matching distribution found for msgpack-python (from PsychoPy==3.0.0b12)
> 
> (anaconda-py3.6) C:\Users\Tim\Documents\Psychopy>conda list msgpack
> \# packages in environment at C:\Users\Tim\Anaconda3\envs\anaconda-py3.6:
> \#
> \# Name                    Version                   Build  Channel
> msgpack-python            0.6.0           py36he980bc4_1000    file:///C:/Users/Tim/Documents/Psychopy/PerceptLabChannel
> 
> (anaconda-py3.6) C:\Users\Tim\Documents\Psychopy>pip show msgpack
> Name: msgpack
> Version: 0.6.0
> Summary: MessagePack (de)serializer.
> Home-page: https://msgpack.org/
> Author: INADA Naoki
> Author-email: songofacandy@gmail.com
> License: Apache 2.0
> Location: c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages
> Requires:
> Required-by:
> 

The `installlog.txt` file looks like:

> Looking in links: wheels
> Processing c:\users\tim\documents\psychopy\psychopy-3.0.0b12.tar.gz
> Requirement already satisfied: requests[security] in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (2.20.1)
> Requirement already satisfied: numpy in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.15.4)
> Requirement already satisfied: scipy in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.1.0)
> Requirement already satisfied: matplotlib in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.0.2)
> Requirement already satisfied: pandas in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (0.23.4)
> Requirement already satisfied: pillow in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (5.3.0)
> Requirement already satisfied: pyglet in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.3.0b1)
> Requirement already satisfied: pygame in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.9.4)
> Requirement already satisfied: configobj in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (5.0.6)
> Requirement already satisfied: pyopengl in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.1.1a1)
> Requirement already satisfied: soundfile in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (0.10.2)
> Requirement already satisfied: sounddevice in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (0.3.12)
> Requirement already satisfied: python-bidi in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (0.4.0)
> Requirement already satisfied: arabic_reshaper in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (2.0.14)
> Requirement already satisfied: cffi in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.11.5)
> Requirement already satisfied: future in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (0.17.0)
> Requirement already satisfied: json_tricks in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.12.2)
> Requirement already satisfied: pyosf in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.0.5)
> Requirement already satisfied: xlrd in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.1.0)
> Requirement already satisfied: openpyxl in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (2.5.12)
> Requirement already satisfied: pyserial in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.4)
> Requirement already satisfied: pyparallel in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (0.2.2)
> Requirement already satisfied: pyyaml in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.13)
> Requirement already satisfied: gevent in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.3.7)
> Collecting msgpack-python (from PsychoPy==3.0.0b12)
> 

I think this may have something to do with versioneer, but I don't know. 

Thanks for any help.

**Comment by tjwilli58** [comment]
If I `pip download --no-cache  msgpack msgpack-python`  to my `wheels` folder, and install it, then when I try to install Psychopy I get:

> 
> (base) C:\Users\Tim\Documents\Psychopy>pip install --no-index -f wheels PsychoPy-3.0.0b12.tar.gz
> Looking in links: wheels
> Processing c:\users\tim\documents\psychopy\psychopy-3.0.0b12.tar.gz
> Requirement already satisfied: requests[security] in c:\users\tim\anaconda3\lib\site-packages (from PsychoPy==3.0.0b12) (2.18.4)
> Requirement already satisfied: numpy in c:\users\tim\anaconda3\lib\site-packages (from PsychoPy==3.0.0b12) (1.14.3)
> Requirement already satisfied: scipy in c:\users\tim\anaconda3\lib\site-packages (from PsychoPy==3.0.0b12) (1.1.0)
> Requirement already satisfied: matplotlib in c:\users\tim\anaconda3\lib\site-packages (from PsychoPy==3.0.0b12) (2.2.2)
> Requirement already satisfied: pandas in c:\users\tim\anaconda3\lib\site-packages (from PsychoPy==3.0.0b12) (0.23.0)
> Requirement already satisfied: pillow in c:\users\tim\anaconda3\lib\site-packages (from PsychoPy==3.0.0b12) (5.1.0)
> Collecting pyglet (from PsychoPy==3.0.0b12)
> Collecting pygame (from PsychoPy==3.0.0b12)
> Collecting configobj (from PsychoPy==3.0.0b12)
>   Could not find a version that satisfies the requirement configobj (from PsychoPy==3.0.0b12) (from versions: )
> No matching distribution found for configobj (from PsychoPy==3.0.0b12)
> 

I think I can just iterate on doing this until everything gets installed.

**Comment by hoechenberger** [comment]
Hi @tjwilli58, we‘re actively working towards providing a proper `conda-forge` package for PsychoPy. Also, just recently three more dependencies are installable via `conda`, see #2122 
You actually don’t need to install _all_ dependencies if you’re not planning to use the specific PsychoPy features that require them. Personally, I _never_ install the full set of dependencies; PsychoPy is installed via `pip install —no-deps` to not pull in all the stuff I’m not going to need anyway. Maybe this could be a viable interim solution for you until a `conda` package becomes available?

**Comment by tjwilli58** [comment]
Thanks @hoechenberger . I didn't know that PsychoPy was installed via --no-deps. I'm working on building my environment with that now.  I was wondering if I could prune the environment-3.6.yml file beyond the one packages I knew I didn't have (x264). I should look at the [manual install instructions](http://psychopy.org/installation.html#manual-install) (is that up-to-date?) to create a minimal environment.yml file.

**Comment by tjwilli58** [comment]
Hi @hoechenberger , I tried `pip --no-deps` to install PsychoPy. No dice:

> (anaconda-py3.6) C:\Users\Tim\Documents\Psychopy>pip install --no-deps --no-index -f wheels PsychoPy-3.0.0b12.tar.gz
> Looking in links: wheels
> Processing c:\users\tim\documents\psychopy\psychopy-3.0.0b12.tar.gz
> Building wheels for collected packages: PsychoPy
>   Running setup.py bdist_wheel for PsychoPy ... done
>   Stored in directory: C:\Users\Tim\AppData\Local\pip\Cache\wheels\b6\89\5a\b9298e95c2f58cef594f740bfb3fe4c989149872a4d0a18554
> Successfully built PsychoPy
> Installing collected packages: PsychoPy
>   Found existing installation: PsychoPy 3.0.0b12
>     Uninstalling PsychoPy-3.0.0b12:
>       Successfully uninstalled PsychoPy-3.0.0b12
> Successfully installed PsychoPy-3.0.0b12
> 
> (anaconda-py3.6) C:\Users\Tim\Documents\Psychopy>python -m psychopy.app.psychopyApp
> Traceback (most recent call last):
>   File "C:\Users\Tim\Anaconda3\envs\anaconda-py3.6\lib\runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "C:\Users\Tim\Anaconda3\envs\anaconda-py3.6\lib\runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "C:\Users\Tim\Anaconda3\envs\anaconda-py3.6\lib\site-packages\psychopy\app\psychopyApp.py", line 100, in <module>
>     start_app()
>   File "C:\Users\Tim\Anaconda3\envs\anaconda-py3.6\lib\site-packages\psychopy\app\psychopyApp.py", line 22, in start_app
>     from psychopy.app._psychopyApp import PsychoPyApp
>   File "C:\Users\Tim\Anaconda3\envs\anaconda-py3.6\lib\site-packages\psychopy\app\_psychopyApp.py", line 51, in <module>
>     from psychopy import projects
>   File "C:\Users\Tim\Anaconda3\envs\anaconda-py3.6\lib\site-packages\psychopy\projects\__init__.py", line 11, in <module>
>     from . import pavlovia
>   File "C:\Users\Tim\Anaconda3\envs\anaconda-py3.6\lib\site-packages\psychopy\projects\pavlovia.py", line 28, in <module>
>     import gitlab
> ModuleNotFoundError: No module named 'gitlab'
> 

When I try to install with dependencies:

> (anaconda-py3.6) C:\Users\Tim\Documents\Psychopy>pip install  --no-index -f wheels PsychoPy-3.0.0b12.tar.gz
> Looking in links: wheels
> Processing c:\users\tim\documents\psychopy\psychopy-3.0.0b12.tar.gz
> Requirement already satisfied: requests[security] in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (2.20.1)
> Requirement already satisfied: numpy in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.15.4)
> Requirement already satisfied: scipy in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.1.0)
> Requirement already satisfied: matplotlib in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.0.2)
> Requirement already satisfied: pandas in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (0.23.4)
> Requirement already satisfied: pillow in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (5.3.0)
> Collecting pyglet (from PsychoPy==3.0.0b12)
> Collecting pygame (from PsychoPy==3.0.0b12)
> Requirement already satisfied: configobj in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (5.0.6)
> Requirement already satisfied: pyopengl in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.1.1a1)
> Collecting soundfile (from PsychoPy==3.0.0b12)
> Collecting sounddevice (from PsychoPy==3.0.0b12)
> Requirement already satisfied: python-bidi in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (0.4.0)
> Collecting arabic_reshaper (from PsychoPy==3.0.0b12)
> Requirement already satisfied: cffi in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.11.5)
> Requirement already satisfied: future in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (0.17.0)
> Requirement already satisfied: json_tricks in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.12.2)
> Requirement already satisfied: pyosf in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.0.5)
> Requirement already satisfied: xlrd in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.1.0)
> Requirement already satisfied: openpyxl in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (2.5.12)
> Requirement already satisfied: pyserial in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.4)
> Collecting pyparallel (from PsychoPy==3.0.0b12)
> Requirement already satisfied: pyyaml in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (3.13)
> Requirement already satisfied: gevent in c:\users\tim\anaconda3\envs\anaconda-py3.6\lib\site-packages (from PsychoPy==3.0.0b12) (1.3.7)
> Collecting msgpack-python (from PsychoPy==3.0.0b12)
>   Could not find a version that satisfies the requirement msgpack-python (from PsychoPy==3.0.0b12) (from versions: )
> No matching distribution found for msgpack-python (from PsychoPy==3.0.0b12)
> 

It looks like I can do a `pip install msgpack-python` to get a little further, then continue doing this until all of the dependencies are found. This will build my "wheels" folder.

**Comment by hoechenberger** [comment]
> Thanks @hoechenberger . I didn't know that PsychoPy was installed via --no-deps. 

Well that’s just how _I_ do it. It’s not the recommended way, and it’s not user-friendly either. I do the `no-deps` install and then try to run PsychoPy, identifying the packages _I_ need to install through the generated `ImportError`s.

> I'm working on building my environment with that now. I was wondering if I could prune the environment-3.6.yml file beyond the one packages I knew I didn't have (x264). 

Yes I think you could and you should. I think the environment files we currently use for Travis also still contain some secondary dependencies (deps of deps) that should ideally be removed.

**Comment by tjwilli58** [comment]
I started looking at the manual install instructions, commenting out the lines in the *.yml file that weren't in the REQUIRED, alternative audio, excel file reading, and iohub sections. It's amazing how much I commented out!

I then looked at the `install requires` section in `setup.cfg`.  Almost all of it was in the `environment-3.6.yml` file, I'm guessing because the *.yml file is generated from that.

I'll play around with an environment file that works for me, but I'm a bit torn on taking out modules that I know won't be used on an offline system, e.g.

> \# making online connections (e.g. OSF.io)
> pip install requests[security] pyosf
> 
or other modules I may need that I don't realize yet.  I suppose I could always modify my environment later though.
